src/core/lib/iomgr/combiner.cc

```diff
@@ -228,29 +211,20 @@ bool grpc_combiner_continue_exec_ctx() {
                               grpc_core::ExecCtx::Get()->IsReadyToFinish(),
                               lock->time_to_execute_final_list));
 
-  if (contended && grpc_core::ExecCtx::Get()->IsReadyToFinish() &&
-      grpc_executor_is_threaded()) {
-    GPR_TIMER_MARK("offload_from_finished_exec_ctx", 0);
-    // this execution context wants to move on: schedule remaining work to be
-    // picked up on the executor
-    queue_offload(lock);
-    return true;
-  }
-
   if (!lock->time_to_execute_final_list ||
       // peek to see if something new has shown up, and execute that with
       // priority
       (gpr_atm_acq_load(&lock->state) >> 1) > 1) {
-    gpr_mpscq_node* n = gpr_mpscq_pop(&lock->queue);
-    GRPC_COMBINER_TRACE(
-        gpr_log(GPR_INFO, "C:%p maybe_finish_one n=%p", lock, n));
-    if (n == nullptr) {
-      // queue is in an inconsistent state: use this as a cue that we should
-      // go off and do something else for a while (and come back later)
-      GPR_TIMER_MARK("delay_busy", 0);
-      queue_offload(lock);
+    gpr_mpscq_node* n = nullptr;
+    bool empty = false;
+    while (!empty && n == nullptr) {
```

Comment:
do-while maybe?

Reply:
I'm hoping it would not get stuck in this loop. If it is a rare scenario, it may be a valid case for a log statement. I get that it is implemented in a way that the inconsistent state would go away soon.

Up to you.